mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+        # (should be the case in correct preprocessing of submission)
+
+        # acl
+        if not ds_md["mdf"].get("acl"):
+            ds_md["mdf"]["acl"] = self.__base_acl or ["public"]
+
+        # version
+        if not ds_md["mdf"].get("version"):
+            ds_md["mdf"]["version"] = 1
+
+        # BLOCK: mrr
+        # TODO
+
+        # Services
+        ds_md["services"] = ds_md.get("services", {})
+
+        # Data
+        ds_md["data"] = ds_md.get("data", {})
+
+        # BLOCK: custom
+        # Make all values into strings
+        if ds_md.get("custom"):
+            new_custom = {}
+            for key, val in ds_md["custom"].items():
+                new_custom[key] = str(val)
+            ds_md["custom"] = new_custom
+
+        # Require strict JSON
+        try:
+            json.dumps(ds_md, allow_nan=False)
+        except (ValueError, json.JSONDecodeError) as e:
+            return {
+                "success": False,
+                "error": "Invalid dataset JSON: {}".format(str(e)),
+                "details": repr(e)
+            }
+
+        # Remove null/None values
+        ds_md = _remove_nulls(ds_md, self.__allowed_nulls)
+
+        # Validate against schema
+        try:
+            jsonschema.validate(ds_md, schema, resolver=resolver)
+        except jsonschema.ValidationError as e:
+            return {
+                "success": False,
+                "error": "Invalid dataset metadata: " + str(e).split("\n")[0],
+                "details": str(e)
+            }
+
+        # Check projects blocks allowed
+        # If no blocks, disallow projects
+        if not self.__project_blocks:
+            if ds_md.get("projects"):
+                return {
+                    "success": False,
+                    "error": "Unauthorized project metadata: No projects allowed",
+                    "details": "'project' block not allowed: '{}'".format(ds_md)
+                }
+        # If some project blocks allowed, check that only allowed ones are present
+        else:
+            unauthorized = []
+            for proj in ds_md.get("projects", {}).keys():
+                if proj not in self.__project_blocks:
+                    unauthorized.append(proj)
+            if unauthorized:
+                return {
+                    "success": False,
+                    "error": ("Unauthorized project metadata: '{}' not allowed"
+                              .format(unauthorized)),
+                    "details": ("Not authorized for project block(s) '{}' in '{}'. "
+                                "The dataset is not in an allowed organization."
+                                .format(unauthorized, ds_md))
+                }
+
+        # Validate required fields
+        # TODO: How should this validation be done?
+        # The metadata conforms to the schema, there are just extra
+        # `requires` values. Perhaps add these to the schema instead?
```

Comment:
I support this. What we have here seems like a challenge that is solved by the JSON schema and it's Python validator.

Reply:
It's also a challenge to correctly monkeypatch the schema. For context, the extra "requires" values are dynamic, and set by the user (or, more commonly, set by options chosen by the user) during dataset submission. It's not possible to bake these "requires" into the base schema; they would have to be added at runtime.
Do you have thoughts on how to easily add the extra "requires" to the correct places in the preset JSONSchema? (I've been bitten trying to do that before; JSONSchema is just verbose enough that it's difficult to programmatically traverse dynamically, without errors.)

Reply:
Ah, thanks for clarifying! 

No, I don't have any ideas on how to traverse the schema. Maybe you could write out JMESPath specs on the fields you want to check and delegate recursing into the records to that code? 

That said, this works and you've already marked this as a "semi-temporary" measure

Reply:
Using JMESPath could be an enhancement in the future, but for now since this code is functional I think we can punt it down the road.